rosidl_typesupport_opendds_c/resource/srv__type_support_c.cpp.em

```diff
@@ -132,71 +180,187 @@ static void * create_replier__@(service.namespaced_type.name)(
   void * untyped_ros_response)
 {
-@# TODO: Implement, considering original code in ffe10f9 or earlier
-  return true;
+  using RequesterType = rosidl_typesupport_opendds_cpp::Requester<
+    @(__dds_msg_typesupport_type)RequestWrapper,
+    @(__dds_msg_typesupport_type)ResponseWrapper
+  >;
+  if (!untyped_requester || !request_header || !untyped_ros_response) {
+    return false;
+  }
+
+  RequesterType * requester = static_cast<RequesterType *>(untyped_requester);
+  @(__dds_msg_typesupport_type)ResponseWrapper dds_response_wrapper;
+
+  if (DDS::RETCODE_OK != requester->take_reply(dds_response_wrapper)) {
+    fprintf(stderr, "take_reply failed\n");
+    return false;
+  }
+
+  int64_t sequence_number =
+    (((int64_t)dds_response_wrapper.header().related_request_id().sequence_number().high) << 32) |
+    dds_response_wrapper.header().related_request_id().sequence_number().low;
+  request_header->sequence_number = sequence_number;
+
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(
+    rosidl_typesupport_opendds_c,
+    @(', '.join(service.response_message.structure.namespaced_type.namespaces)),
+    @(service.response_message.structure.namespaced_type.name))();
+  const message_type_support_callbacks_t * callbacks =
+    static_cast<const message_type_support_callbacks_t *>(ts->data);
+  bool converted = callbacks->convert_dds_to_ros(&(dds_response_wrapper.response()), untyped_ros_response);
+
+  return converted;
 }
 
 bool send_response__@(service.namespaced_type.name)(
   void * untyped_replier,
   const rmw_request_id_t * request_header,
   const void * untyped_ros_response)
 {
-@# TODO: Implement, considering original code in ffe10f9 or earlier
+  using DDSResponseType = @(__dds_response_msg_type);
+  using ReplierType = rosidl_typesupport_opendds_cpp::Replier<
+    @(__dds_msg_typesupport_type)RequestWrapper,
+    @(__dds_msg_typesupport_type)ResponseWrapper
+  >;
+  if (!untyped_replier || !request_header || !untyped_ros_response) {
+    return false;
+  }
+
+  DDSResponseType dds_response;
+
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(
+    rosidl_typesupport_opendds_c,
+    @(', '.join(service.response_message.structure.namespaced_type.namespaces)),
+    @(service.response_message.structure.namespaced_type.name))();
+  const message_type_support_callbacks_t * callbacks =
+    static_cast<const message_type_support_callbacks_t *>(ts->data);
+  bool converted = callbacks->convert_ros_to_dds(untyped_ros_response, &dds_response);
+
+  if (!converted) {
+   fprintf(stderr, "failed to convert ROS response to DDS\n");
+   return false;
+  }
+
+  @(__dds_msg_typesupport_type)ResponseWrapper response_wrapper;
+  response_wrapper.response(dds_response);
+
+  @# Convert request_header to related_request_id
+  @(__rpc_header_prefix)SampleIdentity related_request_id;
+  OpenDDS::DCPS::RepoId id;
+  size_t IDENTITY_SIZE = 16;
```

Comment:
Same as previous comment. Where does the literal 16 come from? Is there a programmatic way to determine this value or can we reference a spec in a comment? Also how do we know that the destination is sufficiently allocated at this point to consume 16 of `&(request_header->writer_guid[0])`